### PR TITLE
AP_Scripting/AP_Mount: add support for DJIR2/R3 handheld gimbals

### PR DESCRIPTION
**Pull request by rmackay9**
This adds support for the DJIR2 and R3 handheld gimbals via a Lua script.

- AP_Mount_Scripting backend added that hold the latest position, angle and/or rate targets from the pilot or mission.  These targets are accessible to scripting using these accessors
  - get_location_target returns the location target if available which is useful for gimbals that accept location targets
  - get_angle_target returns the latest roll, pitch and yaw angles
  - get_rate_target return the latest roll, pitch, yaw rates which are normally provided by the pilot's RC input
  - set_attitude_euler allows the script to pass the actual lean angles from the gimbal to AP for reporting
  - get_camera_state allows the caller to check if it should trigger taking a picture, start/stop recording video, zoom or focus
- AP_Scripting bindings added for the above accessors
- mount-djir2-driver.lua (and .md) that allows control of a DJIR2 or R3 gimbal (wiki page coming shortly).
  - CAN packets are sent to the gimbal using FrameId 0x223, replies appear with FrameId 0x222
  - The DJI R SDK protocol uses serial-over-CAN meaning that it is really a serial protocol with a predefined format including a header, length, two CRCs etc but the serial bytes are packed up into CAN packets of no more than 8 bytes each.  The Lua scripts send_msg() and read_incoming_packets() are responsible for this simple CAN<->Serial conversion.
  - The script requests the gimbal's current attitude at 10hz and also sends the target angle or rate at 10hz

References
- Wiki PR https://github.com/ArduPilot/ardupilot_wiki/pull/4912
- Issue https://github.com/ArduPilot/ardupilot/issues/3107
- Discussion https://discuss.ardupilot.org/t/ap-support-for-dji-rs2-rsc2-rs3-gimbals-maybe/91830
- [DJI R SDK datasheet](https://www.dropbox.com/s/rkmgqecmp12nr2a/DJI_R_SDK_Protocol_and_User_Interface_EN_v2.5.pdf?dl=0)
- ConstantRobotics DJI R SDK https://github.com/ConstantRobotics/DJIR_SDK
- [ROS Serial<->CAN encoder/decoder python script](https://github.com/ceinem/dji_rs2_ros_controller/blob/main/dji_rs2_ros_controller/scripts/can_bus_controller.py)

This has been tested on real hardware and appears to be working correctly.

Special thanks to [Magnus](https://discuss.ardupilot.org/u/magnus/summary) who made his gimbal available for remote testing and development and also to [Julled ](https://discuss.ardupilot.org/u/julled)who provided a key piece of information on the CAN protocol being used.

### CODE REVIEW

**Comment by booo** [comment]
I guess you did not develop this in a clean room so it would be appropriate to mention/attribute the sources you used to develop this somewhere.

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+    0x9601, 0x56c0, 0x5780, 0x9741, 0x5500, 0x95c1, 0x9481, 0x5440,
+    0x9c01, 0x5cc0, 0x5d80, 0x9d41, 0x5f00, 0x9fc1, 0x9e81, 0x5e40,
+    0x5a00, 0x9ac1, 0x9b81, 0x5b40, 0x9901, 0x59c0, 0x5880, 0x9841,
+    0x8801, 0x48c0, 0x4980, 0x8941, 0x4b00, 0x8bc1, 0x8a81, 0x4a40,
+    0x4e00, 0x8ec1, 0x8f81, 0x4f40, 0x8d01, 0x4dc0, 0x4c80, 0x8c41,
+    0x4400, 0x84c1, 0x8581, 0x4540, 0x8701, 0x47c0, 0x4680, 0x8641,
+    0x8201, 0x42c0, 0x4380, 0x8341, 0x4100, 0x81c1, 0x8081, 0x4040
+}
+
+local crc32_lookup = {
```
Comment (magicrub) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:165]: this matches https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Math/crc.cpp#L216

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# DJIRS2 Mount Driver
+
+DJIRS2 gimbal mount driver lua script
+
+How to use
+  Connect gimbal to autopilot's CAN1 port or CAN2 port
```
Comment (tridge) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.md:6]: missing - at start

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# DJIRS2 Mount Driver
+
+DJIRS2 gimbal mount driver lua script
+
+How to use
```
Comment (tridge) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.md:5]: should be ##

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua [hunk 4]
```diff
@@ -0,0 +1,734 @@
+-- mount-djirs2-driver.lua: DJIRS2 mount/gimbal driver
```
Comment (tridge) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:1]: block comment is:
```
--[[
blah1
blah2
--]]
```

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+    crc = ((crc >> 8) & 0xFF) ~ lookup_val
+  end
+  return crc
+end
+
+-- calculate crc32 for a series of bytes
+-- byte_array should be a table of uint8 values
+-- start_byte should be the first byte to start from (using 1 indexing) or left as nil to default to the first byte
+-- num_bytes should be the number of bytes to process or left as nil to use the entire message
+function calc_crc32(byte_array, start_byte, num_bytes)
```
Comment (magicrub) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:221]: This whole chunk of code could be replaced with a simple call to crc_crc32() and give it 0x3AA3 as an initial value

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+local last_print_ms = 0                 -- system time that debug output was last printed
+local bytes_read = 0                    -- number of bytes read from gimbal
+local bytes_written = 0                 -- number of bytes written to gimbal
+local bytes_error = 0                   -- number of bytes read that could not be parsed
+local msg_ignored = 0                   -- number of ignored messages (because frame id does not match)
+local write_fails = 0                   -- number of times write failed
+local execute_fails = 0                 -- number of times that gimbal was unable to execute the command
+local reply_timeouts = 0                -- number of timeouts waiting for replies
+
+local crc16_lookup = {
```
Comment (magicrub) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:130]: where did his 16bit crc table come from?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua (hunk 6 above)
Comment (magicrub) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:130]: image

Turns out this uses our crc_crc16_ibm() with a different seed. Take a look at this:
https://www.lddgo.net/en/encrypt/crc

Comment (hendjoshsr71) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:130]: This one is this according to this tool here,
http://www.sunshine2k.de/coding/javascript/crc/crc_js.html

image

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua (hunk 4 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:1]: txs, replaced individual line comments with block comment.

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua (hunk 6 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:130]: Thanks for this.  I will replace the crc methods with bindings to AP's existing crc functions in a follow-up PR.

**Comment by rmackay9** [comment]
@booo,

txs for the feedback.  I've added references to the two code sources used.

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.md (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.md]: added ##

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.md (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.md]: added "-" before comments

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua (hunk 5 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:228]: cool txs.  I'll move to use the C++ crc functions in a follow-up PR.

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:172]: cool txs.  I'll move to use the C++ crc functions in a follow-up PR.

**Pull request merged**